synapse/cortex.py

```diff
@@ -2061,51 +2058,43 @@ def getNidNdef(self, nid):
             return s_msgpack.un(byts)
 
     def setNidNdef(self, nid, ndef):
+        buid = s_common.buid(ndef)
+        self.v3stor.put(nid, buid, db=self.nid2buid)
+        self.v3stor.put(buid, nid, db=self.buid2nid)
         self.v3stor.put(nid, s_msgpack.en(ndef), db=self.nid2ndef)
 
+        if (nid := int.from_bytes(nid)) >= self.nextnid:
+            self.nextnid = nid + 1
+
     def getBuidByNid(self, nid):
         return self.v3stor.get(nid, db=self.nid2buid)
 
     def getNidByBuid(self, buid):
         return self.v3stor.get(buid, db=self.buid2nid)
 
-    def genBuidNid(self, buid, inc=1):
+    async def genNdefNid(self, ndef):
+        buid = s_common.buid(ndef)
+        nid = self.v3stor.get(buid, db=self.buid2nid)
+        if nid is not None:
+            return nid
+        return await self._push('nid:gen', ndef)
 
+    @s_nexus.Pusher.onPush('nid:gen')
+    async def _genNdefNid(self, ndef):
+        buid = s_common.buid(ndef)
         nid = self.v3stor.get(buid, db=self.buid2nid)
         if nid is not None:
-            refsbyts = self.v3stor.get(nid, db=self.nidrefs)
-            refs = int.from_bytes(refsbyts) + inc
-            self.v3stor.put(nid, refs.to_bytes(8), db=self.nidrefs)
             return nid
 
         nid = self.nextnid.to_bytes(8)
         self.nextnid += 1
 
         self.v3stor.put(nid, buid, db=self.nid2buid)
-        self.v3stor.put(nid, inc.to_bytes(8), db=self.nidrefs)
-
         self.v3stor.put(buid, nid, db=self.buid2nid)
+        self.v3stor.put(nid, s_msgpack.en(ndef), db=self.nid2ndef)
 
```

Comment:
think this needs to go above the `self.v3stor.put(buid, nid, db=self.buid2nid)` line